test/e2e/fishapp/dynamic_stack.go

```diff
@@ -0,0 +1,811 @@
+package fishapp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	appmeshv1beta1 "github.com/aws/aws-app-mesh-controller-for-k8s/pkg/apis/appmesh/v1beta1"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/fishapp/shared"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/collection"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/k8s"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/test/e2e/framework/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+const (
+	connectivityCheckRate = time.Second / 100
+	appMeshOperationRate  = time.Second * 2
+)
+
+// A dynamic generated stack designed to test app mesh integration :D
+// Suppose we have configuration:
+//		5 VirtualServicesCount
+//		10 VirtualNodesCount
+//		2 RoutesCountPerVirtualRouter
+//		2 TargetsCountPerRoute
+//		4 BackendsCountPerVirtualNode
+// we will generate below virtual service configuration & virtual node configuration
+// =======virtual services =========
+//	vs1 -> /path1 -> vn1(50)
+//				  -> vn2(50)
+//		-> /path2 -> vn3(50)
+//				  -> vn4(50)
+//	vs2 -> /path1 -> vn5(50)
+//				  -> vn6(50)
+//		-> /path2 -> vn7(50)
+//				  -> vn8(50)
+//	vs3 -> /path1 -> vn9(50)
+//				  -> vn10(50)
+//		-> /path2 -> vn1(50)
+//				  -> vn2(50)
+//	vs4 -> /path1 -> vn3(50)
+//				  -> vn4(50)
+//		-> /path2 -> vn5(50)
+//				  -> vn6(50)
+//	vs5 -> /path1 -> vn7(50)
+//				  -> vn8(50)
+//		-> /path2 -> vn9(50)
+//				  -> vn10(50)
+// =======virtual nodes =========
+//  vn1 -> vs1,vs2,vs3,vs4
+//  vn2 -> vs5,vs1,vs2,vs3
+//  vn3 -> vs4,vs5,vs1,vs2
```

Comment:
I have some concerns with structuring test fixtures in this way.

First, the mesh does not represent anything corresponding to a real-world use case. Because of this, it is difficult for readers of the test code to reason about. Having a real-world scenario described in a test fixture enables readers to reasonably understand and explain test assertions.

Secondly, you are always creating a symmetrical mesh. In other words, the "dynamic" stack is designed so that it actually *always* produces an mesh where each virtual service is constructed in an identical manner. This will inevitably lead to a greater chance of false positive test results that tend to exclude edge/corner cases that materialize when you have scenarios that are "uneven" (i.e. more real-worldish ;) ).

As an example of how this works out, let me take this assertion function example from this patch:

```go
// expects connectivity and routing works correctly
func (s *DynamicStack) ExpectBehavesCorrectly(ctx context.Context, f *framework.Framework) {
	time.Sleep(3 * time.Minute)
	vsByName := make(map[string]*appmeshv1beta1.VirtualService)
	for i := 0; i != s.VirtualServicesCount; i++ {
		vs := s.createdServiceVSs[i]
		vsByName[vs.Name] = vs
	}

	podsConnectivityCheckResult := make(map[string]map[string]map[string]int)
	for i := 0; i != s.VirtualNodesCount; i++ {
		dp := s.createdNodeDPs[i]
		vn := s.createdNodeVNs[i]

		var vsList []*appmeshv1beta1.VirtualService
		for _, backend := range vn.Spec.Backends {
			vsList = append(vsList, vsByName[backend.VirtualService.VirtualServiceName])
		}
		ret := s.checkDeploymentVirtualServiceConnectivity(ctx, f, dp, vsList)
		for k, v := range ret {
			podsConnectivityCheckResult[k] = v
		}
	}
	prettyJSON, err := json.MarshalIndent(podsConnectivityCheckResult, "", "    ")
	Expect(err).NotTo(HaveOccurred())
	utils.Logf("%v", string(prettyJSON))
}
```

Now, when I read the above, I don't get any semantic or scenario-specific context at all. I don't really know **why** a particular virtual node or virtual service is expected to exist or not exist. It's just too abstracted from reality.

If instead of defining a generic always-symmetric mesh with abstract node and service names, imagine creating **named** real-world mesh examples with realistic names:

```go
type MeshFixture interface {
    Name() string
    Deploy() error
    Check() error
    Cleanup() error
}

// A blue-green deployment scenario, implements MeshFixture
type BlueGreenScenario struct {
    webapp *corev1.Deployment
    service *corev1.Service
    virtService *appmeshv1beta1.VirtualService
    routeConfig shared.RouteToWeightedVirtualNodes
    blueNode *appmeshv1beta1.VirtualNode
    greenNode *appmeshv1beta1.VirtualNode
}

func (d *BlueGreenScenario) Deploy() error {
    // Create the Deployment, Service, VirtualService, Mesh and VirtualNode objects
    // NOTE: do NOT create the VirtualRouter yet
}

func (d *BlueGreenScenario) Check() error {
    // Assert that the Mesh, VirtualNodes, VirtualService, Deployment and Service objects exist properly
    ...
    // Create the VirtualRouter which establishes the weighted routing of traffic to
    // the Green virtual node
    ...
    // Check that traffic is flowing to the Green virtual node now
   ...
}
```

The individual assertions within the `Check()` method will, by nature of this setup, be much more descriptive. Instead of failure messages looking like this:

```go
	Expect(len(pods.Items)).NotTo(BeZero())
```
```
check connectivity of pod 1-a2ka to service 1
Expected 0 to not be 0
```

to instead be something like this:

```
	Expect(len(d.webapp.Pods)).NotTo(BeZero())
```
```
        checking connectivity of pod green-a2ka to service web-app
```

Reply:
Hi Jay, I agree this dynamic test case is not enough.

The the whole point of this PR is to create a test framework and a test case to cover the basic functionalities of AppMesh(multiple service/multiple node/multiple route/multiple target/multiple backend). **It's not intended to cover all E2E scenarios given the timeline i have.**
The dynamic stack is designed in a way to make it reusable for scale tests as well, which allows to do scale test with any dimensions.

There are more tests cases should be covered, e.g. add/remove route, change weights, etc.

The test framework is structured in a way to add new test cases easily. 
e.g. 
* the folder *test/e2e/fishapp* is specific to this `colorApp+httpProxy` container, it can be used to test any http-based appMesh scenarios. 
* the file *test/e2e/fishapp/dynamic_stack is specific to this symmetric mesh based on "fishapp".
* [**extension point**] more tests based on this "fishapp" container(which can be used as either client/server or both at same time)
* [**extension point**] other test scenario not rely on "fishapp" can be added like `test/e2e/color-app`, `test/e2e/bluegreen-app`, etc...

BTW, i didn't design a MeshFixture like interface because these tests tends to be highly depends on use case, some scenario based tests may have deploy & test steps interleaved, there is no value added to have such interface.

Reply:
> But the purpose of this PR is to create a test framework and a test case to cover the basic functionalities of AppMesh(multiple service/multiple node/multiple route/multiple target/multiple backend). **It's not intended to cover all E2E scenarios given the timeline i have.**

I didn't say that I expected you to cover all possible e2e scenarios :)

> The dynamic stack is designed in a way to make it reusable for scale tests as well, which allows to do scale test with configurable dimensions. (I found an appmesh dataplane bug with it )

Yes, I understand that, and the dynamic stack fixture you've created here is a good way to do fuzz, edge/negative, and stress testing. No problem with that.

What I would like to see in a followup PR is an approach that covers the standard use cases/scenarios that AppMesh is designed to cover (blue/green, A/B, etc) in a manner that is easier for readers of the test cases to figure out expectations. Again, it's cool what you've done here. I just think a future enhancement to simplify the scenarios would be useful.

> There are more tests cases yet to be covered, e.g. add/remove route, change weights, http2, grpc, tls, retry, etc.

Yes, agreed completely.

> The test framework is structured in a way to add new test cases easily.
> e.g.
> 
> * the folder _test/e2e/fishapp_ is specific to this `colorApp+httpProxy` container, it can be used to test any http-based appMesh scenarios.
> * the file *test/e2e/fishapp/dynamic_stack is specific to this symmetric mesh based on "fishapp".
> * [**extension point**] more tests based on this "fishapp" container(which can be used as either client/server or both at same time), e.g. `test/e2e/fishapp/bluegreen_stack`
> * [**extension point**] other test scenario not rely on "fishapp" can be added like `test/e2e/color-app`, `test/e2e/bluegreen-app`, etc...

Yep, no disagreement on this.

> BTW, i didn't design a MeshFixture like interface because these tests tends to be highly depends on use case, some scenario based tests may have deploy & test steps interleaved, there is no value added to have such interface.

My suggestion is always code to an interface, not a concrete struct. Makes things more extensible and easier to reason about in the long run. :)

Reply:
😄 TBH, my original version of E2E test is using a blue green deployment color-app. But it cannot cover the case of multiple-path and multiple virtual service. So instead I took this dynamic symmetrical mesh approach.  (I was also asked to script my scale tests, so it can be reused xD).

Yeah, we can add more real-word e2e tests into this in the future :D